File: tests/inter-language-test/src/main/cpp/IltConsumerFilteredBroadcastSubscriptionTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "IltAbstractConsumerTest.h"
#include "joynr/ISubscriptionListener.h"
#include "joynr/OnChangeSubscriptionQos.h"
#include "joynr/Semaphore.h"
#include "joynr/SubscriptionListener.h"
#include "joynr/interlanguagetest/TestInterfaceBroadcastWithFilteringBroadcastFilterParameters.h"
#include "joynr/serializer/Serializer.h"

using namespace ::testing;

class IltConsumerFilteredBroadcastSubscriptionTest : public IltAbstractConsumerTest<::testing::Test>
{
public:
    IltConsumerFilteredBroadcastSubscriptionTest() = default;
};

joynr::Logger iltConsumerFilteredBroadcastSubscriptionTestLogger(
        "IltConsumerFilteredBroadcastSubscriptionTest");

class MockBroadcastWithFilteringBroadcastListener
        : public joynr::ISubscriptionListener<
                  std::string,
                  std::vector<std::string>,
                  joynr::interlanguagetest::namedTypeCollection2::
                          ExtendedTypeCollectionEnumerationInTypeCollection::Enum,
                  joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray,
                  std::vector<
                          joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray>>
{
public:
    MOCK_METHOD1(onSubscribed, void(const std::string& subscriptionId));
    MOCK_METHOD5(
            onReceive,
            void(const std::string& stringOut,
                 const std::vector<std::string>& stringArrayOut,
                 const joynr::interlanguagetest::namedTypeCollection2::
                         ExtendedTypeCollectionEnumerationInTypeCollection::Enum& enumerationOut,
                 const joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray&
                         structWithStringArrayOut,
                 const std::vector<
                         joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray>&
                         structWithStringArrayArrayOut));
    MOCK_METHOD1(onError, void(const joynr::exceptions::JoynrRuntimeException& error));
};

TEST_F(IltConsumerFilteredBroadcastSubscriptionTest, callSubscribeBroadcastWithFiltering)
{
    Semaphore publicationSemaphore;
    std::vector<std::string> stringArrayOut;
    joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray structWithStringArrayOut;
    std::vector<joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray>
            structWithStringArrayArrayOut;
    std::string subscriptionId;

    auto mockBroadcastWithFilteringBroadcastListener =
            std::make_shared<MockBroadcastWithFilteringBroadcastListener>();
    EXPECT_CALL(*mockBroadcastWithFilteringBroadcastListener, onError(_))
            .Times(0)
            .WillRepeatedly(ReleaseSemaphore(&publicationSemaphore));
    EXPECT_CALL(*mockBroadcastWithFilteringBroadcastListener,
                onReceive(Eq("fireBroadcast"),
                          _,
                          Eq(joynr::interlanguagetest::namedTypeCollection2::
                                     ExtendedTypeCollectionEnumerationInTypeCollection::
                                             ENUM_2_VALUE_EXTENSION_FOR_TYPECOLLECTION),
                          _,
                          _))
            .Times(1)
            .WillRepeatedly(DoAll(SaveArg<1>(&stringArrayOut),
                                  SaveArg<3>(&structWithStringArrayOut),
                                  SaveArg<4>(&structWithStringArrayArrayOut),
                                  ReleaseSemaphore(&publicationSemaphore)));
    EXPECT_CALL(*mockBroadcastWithFilteringBroadcastListener,
                onReceive(Eq("doNotFireBroadcast"),
                          _,
                          Eq(joynr::interlanguagetest::namedTypeCollection2::
                                     ExtendedTypeCollectionEnumerationInTypeCollection::
                                             ENUM_2_VALUE_EXTENSION_FOR_TYPECOLLECTION),
                          _,
                          _))
            .Times(0)
            .WillRepeatedly(DoAll(SaveArg<1>(&stringArrayOut),
                                  SaveArg<3>(&structWithStringArrayOut),
                                  SaveArg<4>(&structWithStringArrayArrayOut),
                                  ReleaseSemaphore(&publicationSemaphore)));
    typedef std::shared_ptr<ISubscriptionListener<
            std::string,
            std::vector<std::string>,
            joynr::interlanguagetest::namedTypeCollection2::
                    ExtendedTypeCollectionEnumerationInTypeCollection::Enum,
            joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray,
            std::vector<joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray>>>
            listenerType;
    listenerType listener(mockBroadcastWithFilteringBroadcastListener);
    JOYNR_ASSERT_NO_THROW({
        joynr::interlanguagetest::TestInterfaceBroadcastWithFilteringBroadcastFilterParameters
                filterParameters;

        std::string filterStringOfInterest = "fireBroadcast";
        std::vector<std::string> filterStringArrayOfInterest = IltUtil::createStringArray();
        std::string filterStringArrayOfInterestJson(
                joynr::serializer::serializeToJson(filterStringArrayOfInterest));

        joynr::interlanguagetest::namedTypeCollection2::
                ExtendedTypeCollectionEnumerationInTypeCollection::Enum filterEnumOfInterest =
                        joynr::interlanguagetest::namedTypeCollection2::
                                ExtendedTypeCollectionEnumerationInTypeCollection::
                                        ENUM_2_VALUE_EXTENSION_FOR_TYPECOLLECTION;
        std::string filterEnumOfInterestJson(
                "\"" +
                joynr::interlanguagetest::namedTypeCollection2::
                        ExtendedTypeCollectionEnumerationInTypeCollection::getLiteral(
                                filterEnumOfInterest) +
                "\"");

        joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray
                filterStructWithStringArray = IltUtil::createStructWithStringArray();
        std::string filterStructWithStringArrayJson(
                joynr::serializer::serializeToJson(filterStructWithStringArray));

        std::vector<joynr::interlanguagetest::namedTypeCollection1::StructWithStringArray>
                filterStructWithStringArrayArray = IltUtil::createStructWithStringArrayArray();
        std::string filterStructWithStringArrayArrayJson(
                joynr::serializer::serializeToJson(filterStructWithStringArrayArray));

        filterParameters.setStringOfInterest(filterStringOfInterest);
        filterParameters.setStringArrayOfInterest(filterStringArrayOfInterestJson);
        filterParameters.setEnumerationOfInterest(filterEnumOfInterestJson);
        filterParameters.setStructWithStringArrayOfInterest(filterStructWithStringArrayJson);
        filterParameters.setStructWithStringArrayArrayOfInterest(
                filterStructWithStringArrayArrayJson);

        JOYNR_LOG_INFO(iltConsumerFilteredBroadcastSubscriptionTestLogger,
                       "callSubscribeBroadcastWithFiltering - register subscription");

        int64_t minInterval_ms = 0;
        int64_t validity = 60000;
        int64_t publicationTtl = UnicastSubscriptionQos::DEFAULT_PUBLICATION_TTL_MS();
        auto subscriptionQos = std::make_shared<joynr::OnChangeSubscriptionQos>(
                validity, publicationTtl, minInterval_ms);
        testInterfaceProxy
                ->subscribeToBroadcastWithFilteringBroadcast(
                        filterParameters, listener, subscriptionQos)
                ->get(subscriptionIdFutureTimeoutMs, subscriptionId);

        JOYNR_LOG_INFO(iltConsumerFilteredBroadcastSubscriptionTestLogger,
                       "callSubscribeBroadcastWithFiltering - subscription registered");

        JOYNR_LOG_INFO(iltConsumerFilteredBroadcastSubscriptionTestLogger,
                       "callSubscribeBroadcastWithFiltering - fire broadast \"fireBroadcast\"");
        std::string stringArg = "fireBroadcast";
        testInterfaceProxy->methodToFireBroadcastWithFiltering(stringArg);
        ASSERT_TRUE(publicationSemaphore.waitFor(publicationTimeoutMs));

        EXPECT_TRUE(IltUtil::checkStringArray(stringArrayOut));
        EXPECT_TRUE(IltUtil::checkStructWithStringArray(structWithStringArrayOut));
        EXPECT_TRUE(IltUtil::checkStructWithStringArrayArray(structWithStringArrayArrayOut));

        JOYNR_LOG_INFO(
                iltConsumerFilteredBroadcastSubscriptionTestLogger,
                "callSubscribeBroadcastWithFiltering - fire broadast \"doNotFireBroadcast\"");
        stringArg = "doNotFireBroadcast";
        testInterfaceProxy->methodToFireBroadcastWithFiltering(stringArg);
        ASSERT_FALSE(publicationSemaphore.waitFor(publicationTimeoutMs));

        testInterfaceProxy->unsubscribeFromBroadcastWithFilteringBroadcast(subscriptionId);
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    });
}

File: tests/inter-language-test/inter-language-test-base/src/test/java/io/joynr/test/interlanguage/IltConsumerFilteredBroadcastSubscriptionTest.java
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.test.interlanguage;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Arrays;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.core.JsonProcessingException;

import io.joynr.exceptions.SubscriptionException;
import io.joynr.proxy.Future;
import joynr.OnChangeWithKeepAliveSubscriptionQos;
import joynr.interlanguagetest.TestInterfaceBroadcastInterface.BroadcastWithFilteringBroadcastAdapter;
import joynr.interlanguagetest.TestInterfaceBroadcastInterface.BroadcastWithFilteringBroadcastFilterParameters;
import joynr.interlanguagetest.namedTypeCollection1.StructWithStringArray;
import joynr.interlanguagetest.namedTypeCollection2.ExtendedTypeCollectionEnumerationInTypeCollection;

public class IltConsumerFilteredBroadcastSubscriptionTest extends IltConsumerTest {
    private static final Logger logger = LoggerFactory.getLogger(IltConsumerFilteredBroadcastSubscriptionTest.class);

    @BeforeClass
    public static void setUp() throws Exception {
        logger.info("setUp: Entering");
        setupConsumerRuntime(false);
        logger.info("setUp: Leaving");
    }

    @AfterClass
    public static void tearDown() throws InterruptedException {
        logger.info("tearDown: Entering");
        generalTearDown();
        logger.info("tearDown: Leaving");
    }

    // variables that are to be changed inside callbacks must be declared global
    volatile boolean subscribeBroadcastWithFilteringCallbackDone = false;
    volatile boolean subscribeBroadcastWithFilteringCallbackResult = false;

    @SuppressWarnings("checkstyle:methodlength")
    @Test
    public void callSubscribeBroadcastWithFiltering() {
        Future<String> subscriptionIdFuture;
        String subscriptionId;
        int minIntervalMs = 0;
        int maxIntervalMs = 10000;
        long validityMs = 60000;
        int alertAfterIntervalMs = 20000;
        int publicationTtlMs = 5000;
        OnChangeWithKeepAliveSubscriptionQos subscriptionQos = new OnChangeWithKeepAliveSubscriptionQos().setMinIntervalMs(minIntervalMs)
                                                                                                         .setMaxIntervalMs(maxIntervalMs)
                                                                                                         .setValidityMs(validityMs)
                                                                                                         .setAlertAfterIntervalMs(alertAfterIntervalMs)
                                                                                                         .setPublicationTtlMs(publicationTtlMs);
        logger.info(name.getMethodName());

        try {
            BroadcastWithFilteringBroadcastFilterParameters filterParameters = new BroadcastWithFilteringBroadcastFilterParameters();
            String stringOfInterest = "fireBroadcast";
            filterParameters.setStringOfInterest(stringOfInterest);

            String[] stringArrayOfInterest = { "Hello", "World" };
            String json;
            try {
                logger.info(name.getMethodName() + " - objectMapper is " + objectMapper);
                logger.info(name.getMethodName() + " - objectMapper stringArrayOfInterest " + stringArrayOfInterest);
                json = objectMapper.writeValueAsString(stringArrayOfInterest);
            } catch (JsonProcessingException je) {
                throw new RuntimeException(getExceptionMessage(je, "stringArrayOfInterest"));
            }
            filterParameters.setStringArrayOfInterest(json);

            ExtendedTypeCollectionEnumerationInTypeCollection enumerationOfInterest = ExtendedTypeCollectionEnumerationInTypeCollection.ENUM_2_VALUE_EXTENSION_FOR_TYPECOLLECTION;
            try {
                json = objectMapper.writeValueAsString(enumerationOfInterest);
            } catch (JsonProcessingException je) {
                throw new RuntimeException(getExceptionMessage(je, "enumerationOfInterest"));
            }
            filterParameters.setEnumerationOfInterest(json);

            StructWithStringArray structWithStringArrayOfInterest = IltUtil.createStructWithStringArray();
            try {
                json = objectMapper.writeValueAsString(structWithStringArrayOfInterest);
            } catch (JsonProcessingException je) {
                throw new RuntimeException(getExceptionMessage(je, "structWithStringArrayOfInterest"));
            }
            filterParameters.setStructWithStringArrayOfInterest(json);

            StructWithStringArray[] structWithStringArrayArrayOfInterest = IltUtil.createStructWithStringArrayArray();
            try {
                json = objectMapper.writeValueAsString(structWithStringArrayArrayOfInterest);
            } catch (JsonProcessingException je) {
                throw new RuntimeException(getExceptionMessage(je, "structWithStringArrayArrayOfInterest"));
            }
            filterParameters.setStructWithStringArrayArrayOfInterest(json);

            subscriptionIdFuture = testInterfaceProxy.subscribeToBroadcastWithFilteringBroadcast(new BroadcastWithFilteringBroadcastAdapter() {
                @Override
                public void onReceive(String stringOut,
                                      String[] stringArrayOut,
                                      ExtendedTypeCollectionEnumerationInTypeCollection enumerationOut,
                                      StructWithStringArray structWithStringArrayOut,
                                      StructWithStringArray[] structWithStringArrayArrayOut) {

                    logger.info(name.getMethodName() + " - callback - got broadcast");

                    String[] stringArray = { "Hello", "World" };
                    if (!Arrays.equals(stringArray, stringArrayOut)) {
                        subscribeBroadcastWithFilteringCallbackResult = false;
                    } else if (enumerationOut != ExtendedTypeCollectionEnumerationInTypeCollection.ENUM_2_VALUE_EXTENSION_FOR_TYPECOLLECTION) {
                        logger.info(name.getMethodName() + " - callback - invalid content");
                        subscribeBroadcastWithFilteringCallbackResult = false;
                    } else if (!IltUtil.checkStructWithStringArray(structWithStringArrayOut)) {
                        logger.info(name.getMethodName() + " - callback - invalid content");
                        subscribeBroadcastWithFilteringCallbackResult = false;
                    } else if (!IltUtil.checkStructWithStringArrayArray(structWithStringArrayArrayOut)) {
                        logger.info(name.getMethodName() + " - callback - invalid content");
                        subscribeBroadcastWithFilteringCallbackResult = false;
                    } else {
                        logger.info(name.getMethodName() + " - callback - content OK");
                        subscribeBroadcastWithFilteringCallbackResult = true;
                    }
                    subscribeBroadcastWithFilteringCallbackDone = true;
                }

                @Override
                public void onError(SubscriptionException error) {
                    logger.info(name.getMethodName() + " - callback - error");
                    subscribeBroadcastWithFilteringCallbackResult = false;
                    subscribeBroadcastWithFilteringCallbackDone = true;
                }
            }, subscriptionQos, filterParameters);
            subscriptionId = subscriptionIdFuture.get(10000);
            logger.info(name.getMethodName() + " - subscription successful, subscriptionId = " + subscriptionId);
            logger.info(name.getMethodName() + " - Waiting one second");
            Thread.sleep(1000);
            logger.info(name.getMethodName() + " - Wait done, invoking fire method");
            String stringArg = "fireBroadcast";
            testInterfaceProxy.methodToFireBroadcastWithFiltering(stringArg);
            logger.info(name.getMethodName() + " - fire method invoked");

            // check results from callback; expect to be finished within 1 second
            // should have been called ahead anyway
            if (!subscribeBroadcastWithFilteringCallbackDone) {
                logger.info(name.getMethodName() + " - about to wait for a second for callback");
                Thread.sleep(1000);
                logger.info(name.getMethodName() + " - wait for callback is over");
            } else {
                logger.info(name.getMethodName() + " - callback already done");
            }

            assertTrue(name.getMethodName() + " - FAILED - callback did not get called in time",
                       subscribeBroadcastWithFilteringCallbackDone);
            assertTrue(name.getMethodName()
                    + " - FAILED - callback got called but received unexpected error or publication content",
                       subscribeBroadcastWithFilteringCallbackResult);
            logger.info(name.getMethodName() + " - callback got called and received expected publication");

            // reset counter for 2nd test
            subscribeBroadcastWithFilteringCallbackResult = false;
            subscribeBroadcastWithFilteringCallbackDone = false;

            logger.info(name.getMethodName() + " - invoking fire method with wrong stringArg");
            stringArg = "doNotfireBroadcast";
            testInterfaceProxy.methodToFireBroadcastWithFiltering(stringArg);
            logger.info(name.getMethodName() + " - fire method invoked");

            // check results from callback; expect to be finished within 1 second
            // should have been called ahead anyway
            if (!subscribeBroadcastWithFilteringCallbackDone) {
                logger.info(name.getMethodName() + " - about to wait for a second for callback");
                Thread.sleep(1000);
                logger.info(name.getMethodName() + " - wait for callback is over");
            } else {
                logger.info(name.getMethodName() + " - callback already done");
            }

            assertFalse(name.getMethodName() + " - FAILED - callback got called unexpectedly",
                        subscribeBroadcastWithFilteringCallbackDone);
            logger.info(name.getMethodName() + " - callback did not get called in time (expected)");

            // try to unsubscribe
            try {
                testInterfaceProxy.unsubscribeFromBroadcastWithFilteringBroadcast(subscriptionId);
                logger.info(name.getMethodName() + " - unsubscribe successful");
            } catch (Exception e) {
                fail(name.getMethodName() + " - FAILED - caught unexpected exception on unsubscribe: "
                        + e.getMessage());
            }

            logger.info(name.getMethodName() + " - OK");
        } catch (Exception e) {
            // also catches InterruptedException from Thread.sleep() call
            StringWriter stringWriter = new StringWriter();
            PrintWriter printWriter = new PrintWriter(stringWriter);
            e.printStackTrace(printWriter);
            printWriter.flush();

            fail(name.getMethodName() + " - FAILED - caught unexpected exception: " + stringWriter.toString());
        }
    }

    private String getExceptionMessage(JsonProcessingException exception, String what) {
        return name.getMethodName() + " - FAILED - got exception when serializing " + what + exception.getMessage();
    }
}
